_posts/2024-03-20-kraft-migration.md

```diff
@@ -0,0 +1,196 @@
+The nodes forming the KRaft controller quorum are all configured with the connection to ZooKeeper together with the additional `zookeeper.metadata.migration.enable=true` flag which states the intention to run the migration.
+When the KRaft controllers start, they form a quorum, elect the leader and move in a state where they are waiting for the brokers to register.
+
+![KRaft controller quorum deployed](/assets/images/posts/2024-03-20-kraft-migration-02-kraft-deployed.png)
+
+#### Enabling brokers to run the migration
+
+The next step is to move the brokers to migration mode.
+In order to do so, the brokers' configuration needs to be updated by adding connection details for all nodes of the KRaft controller quorum and enabling the migration with the `zookeeper.metadata.migration.enable=true` flag.
+After the update, the brokers need to be rolled one by one to make such configuration changes effective.
+On restart, the brokers register to the KRaft controller quorum and the migration begins.
+The KRaft controller leader copies all metadata from ZooKeeper to the `__cluster_metadata` topic.
+
+![KRaft migration running](/assets/images/posts/2024-03-20-kraft-migration-03-kraft-migration.png)
+
+While the migration is running, you can verify its status by looking at the log on the KRaft controller leader or by checking the `kafka.controller:type=KafkaController,name=ZkMigrationState` metric.
+When the migration is completed, that is the metric value is `MIGRATION`, the brokers are anyway still running in ZooKeeper mode.
+The KRaft controllers are in charge of handling any requests related to metadata changes within the cluster but they keep sending RPCs to the brokers for metadata updates.
+The metadata are still copied to ZooKeeper and the cluster is working in a so called "dual-write" mode.
+
+![KRaft dual-write](/assets/images/posts/2024-03-20-kraft-migration-04-kraft-dual-write.png)
+
+#### Moving brokers to be full KRaft
+
+The next step is about moving the brokers to be in full KRaft mode and not using ZooKeeper anymore.
+In order to do so, the brokers configuration is updated by removing the connection to ZooKeeper and disabling the migration flag.
+All the brokers are rolled again and, on restart, they are now in full KRaft mode without any connection or usage of ZooKeeper.
+The KRaft controllers are still in "dual-write" mode and any metadata changes are copied to ZooKeeper.
+
+![Brokers only full KRaft](/assets/images/posts/2024-03-20-kraft-migration-05-brokers-kraft.png)
+
+#### Migration finalization
+
+The final step is to reconfigure the KRaft controllers without the connection to ZooKeeper and disabling the migration flag.
+When all the KRaft controllers are rolled, the cluster is working in full KRaft mode and the ZooKeeper ensemble is not used anymore.
+From now on, it is possible to deprovision the ZooKeeper nodes from your environment.
```

Comment:
Rather than `it is possible`, shouldn't it be `you should`? because at this point ZooKeeper is unused and just using resources/costing money.